### ISSUE TITLE
Roslyn fails to enforce definitely assigned rules and cannot emit code for local functions with closures over pattern variables in switch statements

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#15322]
Following code should have error CS0165: Use of unassigned local variable 'x'

```cs
class Program
{
    static void Main(object[] args)
    {
        switch(args[0])
        {
            case string x:
                Foo(); // missing error
                break;
            case int x:
                void Foo() => System.Console.WriteLine(x);
                break;
        }
    }
}
```

Roslyn doesn't detect the error and fails to compile the code, this is VS2017 RC output:
```
csc.exe exited with code 255
```

Note that if you reorder the cases you get correct error i.e.
```cs
class Program
{
    static void Main(object[] args)
    {
        switch(args[0])
        {
            case int x:
                void Foo() => System.Console.WriteLine(x);
                break;
            case string x:
                Foo(); // error CS0165: Use of unassigned local variable 'x'
                break;
        }
    }
}
```

TryRoslyn emits following stacktrace for original code (it may be a bit outdated though):
```
The given key was not present in the dictionary. 
at System.Collections.Generic.Dictionary`2.get_Item(TKey key) 
at Microsoft.CodeAnalysis.CodeGen.LocalSlotManager.GetLocal(ILocalSymbol symbol) in ...\Src\Compilers\Core\Portable\CodeGen\LocalSlotManager.cs:line 148 
at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.GetLocal(LocalSymbol symbol) in ...\Src\Compilers\CSharp\Portable\CodeGen\EmitStatement.cs:line 1387 
at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.GetLocal(BoundLocal localExpression) in ...\Src\Compilers\CSharp\Portable\CodeGen\EmitStatement.cs:line 1382 
at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitLocalAddress(BoundLocal localAccess) in ...\Src\Compilers\CSharp\Portable\CodeGen\EmitAddress.cs:line 167 
at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitAddress(BoundExpression expression, AddressKind addressKind) in ...\Src\Compilers\CSharp\Portable\CodeGen\EmitAddress.cs:line 38 
at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitArgument(BoundExpression argument, RefKind refKind) in ...\Src\Compilers\CSharp\Portable\CodeGen\EmitExpression.cs:line 594 
at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitArguments(ImmutableArray`1 arguments, ImmutableArray`1 parameters) in ...\Src\Compilers\CSharp\Portable\CodeGen\EmitExpression.cs:line 769 
at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitCallExpression(BoundCall call, UseKind useKind) in ...\Src\Compilers\CSharp\Portable\CodeGen\EmitExpression.cs:line 1480 
at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitExpressionCore(BoundExpression expression, Boolean used) in ...\Src\Compilers\CSharp\Portable\CodeGen\EmitExpression.cs:line 92 
at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitExpressionCoreWithStackGuard(BoundExpression expression, Boolean used) in ...\Src\Compilers\CSharp\Portable\CodeGen\EmitExpression.cs:line 72 
at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitExpression(BoundExpression expression, Boolean used) in ...\Src\Compilers\CSharp\Portable\CodeGen\EmitExpression.cs:line 60 
```

ping @agocke

**Comment by agocke** [comment]
@TessenR Thanks for the ping. I'm betting this is the same problem as #15298. I was trying to be a bit too clever ;)

**Comment by TessenR** [comment]
@agocke What about this code? There are no 'use of unassigned variable' errors so I believe it will compile without errors when you merge the pull request. Is it intended?

```cs
class C
{
  void M()
  {
    Assign(); // there is no 'use of unassigned variable' since 'Assign' doesn't read 'x'
    Local();  // is 'x' assigned here?
    int x;
    void Local() => Console.WriteLine(x);
    void Assign() { x = 5; }
  }
}
```

**Comment by TessenR** [comment]
Also note that Roslyn still can't compile this code and 'x' is always assigned before used so there are no 'use of unassigned variable' errors no matter how I reorder the cases.

```cs
class Program
{
    static void Main(string[] args)
    {
        object o = null;
        switch (o)
        {
            case string x:
                Assign();
                Foo();
                break;
            case int x:
                void Assign() { x = 5; }
                void Foo() => System.Console.WriteLine(x);
                break;
        }
    }
}
```

VS 2017 RC
```
"csc.exe" exited with code -2146232797
```

**Comment by agocke** [comment]
@TessenR The first example I believe is by design. As long as the variable is declared before the local function it is declared, but not definitely assigned, for that local function. I'll add that case as a regression test for my PR.

Regarding the second example -- that is valid code, so crashing during compilation is a bug. I'll file a separate bug for that, since it looks like it doesn't have anything to do with definite assignment and has a different root cause.
